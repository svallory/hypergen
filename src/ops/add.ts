import path from 'node:path'
import fs from 'fs-extra'
import chalk from 'chalk'
import type { ActionResult, RenderedAction, RunnerConfig } from '../types.js'
import createResult from './result.js'

const add = async (
  action: RenderedAction,
  args: Record<string, any>,
  { logger, cwd, createPrompter }: RunnerConfig,
): Promise<ActionResult> => {
  const {
    attributes: { to, inject, unless_exists, force, from, skip_if },
  } = action
  const result = createResult('add', to)
  const prompter = createPrompter()
  if (!to || inject) {
    return result('ignored')
  }
  if (!to) {
    return result('missing "to" attribute')
  }

  const absTo = path.resolve(cwd, to)
  const shouldNotOverwrite =
    !force && unless_exists !== undefined && unless_exists === true
  const fileExists = await fs.pathExists(absTo)

  if (shouldNotOverwrite && fileExists) {
    logger.warn(`     skipped: ${to}`)
    return result('skipped')
  }
  if (!process.env.HYPERGEN_OVERWRITE && fileExists && !force) {
    if (
      !(await prompter
        .prompt({
          prefix: '',
          type: 'confirm',
          name: 'overwrite',
          message: chalk.red(`     exists: ${to}. Overwrite? (y/N): `),
        })
        .then(({ overwrite }) => overwrite))
    ) {
      logger.warn(`     skipped: ${to}`)
      return result('skipped')
    }
  }

  const shouldSkip = skip_if === 'true'

  if (shouldSkip) {
    return result('skipped')
  }

  if (from) {
    const from_path = path.join(args.templates, from)
    const file = fs.readFileSync(from_path).toString()
    action.body = file
  }

  if (!args.dry) {
    await fs.ensureDir(path.dirname(absTo))
    await fs.writeFile(absTo, action.body)
  }
  const pathToLog = process.env.HYPERGEN_OUTPUT_ABS_PATH ? absTo : to
  logger.ok(`       ${force ? 'FORCED' : 'added'}: ${pathToLog}`)

  return result('added')
}

export default add
