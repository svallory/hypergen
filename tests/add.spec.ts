import { describe, it, expect } from 'bun:test'
import fs from 'fs-extra'
import path from 'node:path'
import add from '~/ops/add.js'
import Logger from '~/logger.js'
import MockEnquirer from './util/enquirer.js'

describe('add', () => {
  const logger = new Logger(console.log)

  it('should add a new file', async () => {
    const CWD = '/tmp/add-new-file'
    fs.removeSync(CWD)
    const to = path.join(CWD, 'app/hello.js')

    const result = await add(
      {
        attributes: { to: 'app/hello.js' },
        body: 'hello',
      },
      {},
      {
        logger,
        cwd: CWD,
        debug: false,
        helpers: {},
        createPrompter: () => new MockEnquirer(),
      },
    )

    expect(result).toMatchSnapshot({ timing: expect.any(Number) })
    expect(fs.existsSync(to)).toBe(true)
    expect(fs.readFileSync(to).toString()).toBe('hello')
    fs.removeSync(CWD)
  })

  it('should overwrite an existing file when confirmed', async () => {
    const CWD = '/tmp/overwrite-existing-file'
    fs.removeSync(CWD)
    const to = path.join(CWD, 'app/hello.js')
    fs.ensureDirSync(path.dirname(to))
    fs.writeFileSync(to, 'original content')

    const result = await add(
      {
        attributes: { to: 'app/hello.js' },
        body: 'new content',
      },
      {},
      {
        logger,
        cwd: CWD,
        debug: false,
        helpers: {},
        createPrompter: () => new MockEnquirer(),
      },
    )

    expect(result).toMatchSnapshot({ timing: expect.any(Number) })
    expect(fs.readFileSync(to).toString()).toBe('new content')
    fs.removeSync(CWD)
  })
})
