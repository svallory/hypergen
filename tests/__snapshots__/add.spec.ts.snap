// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`add should add a new file 1`] = `
{
  "status": "added",
  "subject": "app/hello.js",
  "timing": Any<Number>,
  "type": "add",
}
`;

exports[`add should overwrite an existing file when confirmed 1`] = `
{
  "status": "added",
  "subject": "app/hello.js",
  "timing": Any<Number>,
  "type": "add",
}
`;
